caramel/scripts/tool.py

```diff
@@ -79,20 +79,29 @@ def error_out(message):
 
 
 def print_list():
-    requests = models.CSR.valid()
+    valid_requests = (
+        models.DBSession.query(
+            models.CSR.id,
+            models.CSR.commonname,
+            models.CSR.sha256sum,
+            func.max(models.Certificate.not_after),
+        )
+        .outerjoin(models.CSR.certificates)
+        .group_by(models.CSR.id)
+        .filter(models.CSR.rejected.is_(False))
+        .order_by(models.CSR.id)
+        .all()
+    )
 
     def unsigned_last(csr):
```

Comment:
I've just cheked the existing datasystem, and I'm prone to say that unsigned last should be removed and dropped.    It's better to sort by ID once the amount of request goes up high, that causes less issues with risk of mis-signing when there are multiple requests for the same thing.


A common case is that a user will fire off many requests in a row, without waiting for any of them to be signed, leading us to a situation of having to reject one or more.

Reply:
https://xkcd.com/1172/